Introduce a gRPC Tagger service

juliogreff opened the pull request:
### What does this PR do?

This PR introduces a gRPC interface that wraps the Tagger and exposes two RPCs: `StreamTags` to watch and stream changes to entities in a "push" model, and `FetchEntity` to pull the current tags for one entity.

### Motivation

This is the first part in an epic that aims to centralize the Tagger into a single agent, instead of having one per agent.

### Describe your test plan

There is an HTTP/JSON gateway already configured in the agent, so the changes here can be tested with `curl` easily:

```
$  curl -H "authorization: Bearer $(cat /etc/datadog-agent/auth_token)" \
   -XPOST -k https://localhost:5001/v1/grpc/tagger/stream_entities

$ curl -H "authorization: Bearer $(cat /etc/datadog-agent/auth_token)" \
   -XPOST -k -H "Content-Type: application/json" \
   --data '{"id":{"prefix":"kubernetes_pod_uid","uid":"d575fb58-82dc-418e-bfb1-aececc9bc507"}}' \
   https://localhost:5001/v1/grpc/tagger/fetch_entity 
```

juliogreff committed:
Introduce a gRPC Tagger service

clamoriniere reviewed (commented):
Looks great!!! 
Only on open question about the lifecycle between the mutexes in the `tagStore` 
small comment and nits.

juliogreff commented on pkg/tagger/tagstore.go:
```diff
@@ -74,11 +80,24 @@ func (s *tagStore) processTagInfo(info *collectors.TagInfo) error {
+		return err
+	}
+
+	s.notifySubscribers(info.Entity, storedTags, eventType)
```
It seems to me that without the lock on the store, events can be sent out of order, and order is important. For example, an update shouldn't be sent after a delete, and without the lock there's no way to guarantee that order. It's also important when a new subscriber is being added, since it needs to read the store, register the channel, and only then it can start receiving events. Without the lock, I don't think we can preserve that order.

On channels being full: yes, it's quite possible, but I don't whether that will become a problem. I just made a change to alleviate that a bit (instead of sending single events, we group them in a list and send that), but even then it's not guaranteed. I'm not sure what the behavior should be in that case, but I'm thinking that whenever the channel is full it should be closed, which would end the stream and signal to the client to start a fresh new one. I'm happy to hear thoughts on that.

clamoriniere commented on cmd/agent/api/pb/api.proto:
```diff
@@ -15,12 +15,18 @@ service Agent {
 
 service AgentSecure {
-    rpc GetTags (TagRequest) returns (TagReply) {
+    rpc StreamTags(StreamTagsRequest) returns (stream StreamTagsResponse) {
```
I'm wondering if we should use more similar name for the 2 service linked the tagger.

```golang 
rpc TaggerStreamEntities(TaggerStreamEntitiesRequest) returns (stream TaggerStreamEntityResponse) {...}

rpc TaggerFetchEntity(TaggerFetchEntityRequest)  return (TaggerEntityResponse) {...}
```

WDYT?

clamoriniere commented on cmd/agent/api/pb/api.proto:
```diff
@@ -15,12 +15,18 @@ service Agent {
         };
-    }
+    };
+    rpc FetchEntity(FetchEntityRequest) returns (Entity) {
```
should we return an `EntityResponse` instead of an `Entity` ?

juliogreff committed:
Benchmark the tagger.tagStore

To ensure that upcoming changes to the tagger have a controlled impact
on performance, this benchmark aims to measure tagStore.processTagInfo's
performance, but also overall throughput to ensure that lock contention
isn't an issue.

juliogreff commented:
I added some benchmarks to ensure we don't let performance degrade too much. Below are the results. Benchmarks were run with `go test -run=NONE -bench=TagStore -count 5 ./pkg/tagger`.

```shell
$ benchstat old.txt new.txt
name                       old time/op  new time/op  delta
TagStoreThroughput-8        307ms ± 2%   322ms ± 2%  +4.70%  (p=0.008 n=5+5)
TagStore_processTagInfo-8   123µs ±12%   126µs ±12%    ~     (p=0.690 n=5+5)
```

`processTagInfo` being run on its own basically shows no change. The overall throughput is a little bit worse (4.7%), but that's expected since there's extra coordination required to notify subscribers.

truthbk commented on cmd/agent/api/pb/api.proto:
```diff
@@ -15,12 +15,18 @@ service Agent {
+    rpc TaggerStreamEntities(StreamTagsRequest) returns (stream StreamTagsResponse) {
         option (google.api.http) = {
-            post: "/v1/grpc/tags"
+            post: "/v1/grpc/stream_tags"
```
I would maybe put these in: `/v1/grpc/tagger/stream_tags` and `/v1/grpc/tagger/fetch_entity`

truthbk replied:
Also, you might want to add some docs (or comments) somewhere describing what would be the expected JSON format of the messages when the grpc-gateway is used. I know this won't apply much to practical use, but it might be helpful for troubleshooting. We're more likely to send a customer a short command-line curl snippet rather than ask them to install `grpcurl`, and so forth.

truthbk replied:
Finally, with regards to the fact the tagger endpoint is going to be secured. Will this always be the case, or do you anticipate scenarios were the tagger might have to be accessed without authentication? 

There are currently two alternatives for scenarios were access to the gRPC (or legacy API) is required from remote locations:
1. shared secret
2. unauthenticated API access (still encrypted though)

If it could be either/or, we might have to think a little bit about how to expose them.

juliogreff replied:
Thanks for the review @truthbk! I applied the changes to the gateway paths as you suggested, and added documentation.

Regarding authentication, I don't think we'll ever want the tagger endpoints to ever be split between authenticated and non-authenticated. It's possible that we might want to remove auth at some point, but if that ever happens we'll move the whole thing (as in, move all the endpoints from `AgentSecure` to `Agent`), to avoid having to do complicated routing and RBAC.

For remote access, we expect a similar mechanism to that of the DCA, through a shared secret mounted at a known location for the clients, which as I understand is what's implemented for the `serverSecure` server.

truthbk reviewed (commented):
Mostly looks great, I've only really reviewed the grpc part and raised a couple of questions, but it definitely looks solid to me. Let me know your thoughts! 🙇

truthbk reviewed (approved):
🚀

clamoriniere reviewed (approved):
Nice work !!!! 💯
